### PR TITLE
AutoRateLimit based on recent karma

### PR DESCRIPTION
**Pull request by Raemon**
This creates a new set of possible conditions for AutoRateLimits:

– recentKarma (total karma gained by the user's past 20 posts and/or comments, sorted by recency)
– recentPostKarma (total karma from past 20 posts)
– recentCommentKarma (total karma from past 20 comments)

– downvoterCount (unique downvoters from the user's past 20 posts and/or comments)
– postDownvoterCount (unique downvoters from the user's past 20 posts)
– commentDownvoterCount (unique downvoters from the user's past 20 comments)

All are accessible via the Users.recentKarmaInfo resolver field.

The intention here is to (typically) only apply karma thresholds if there are at least a N downvoters (I'm not yet sure what number N should be, but pretty confident the number should be at least 2). 

I'm also not sure whether it'll make more sense to rate limit based on total karma, or to, say, rate limit commenting based on comment karma and posts based on post karma (I think there are at least some users who seem like fine commenters but bad posters, and could imagine vice-versa).

We also might want "woah buddy slow down" rate-limits that, say, limit you to one comment per hour if you've recently got lots of unique downvoters and some-but-not-much karma. (though I'd have to see what patterns tend to be common in demon-threads before having a sense of how to calibrate that)

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1204782304036620) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/components/sunshineDashboard/UsersReviewInfoCard.tsx [hunk 1]
```diff
@@ -7,6 +7,9 @@ import { userCanDo } from '../../lib/vulcan-users/permissions';
 import classNames from 'classnames';
 import { hideScrollBars } from '../../themes/styleUtils';
 import { getReasonForReview } from '../../lib/collections/moderatorActions/helpers';
+import VotesRepo from '../../server/repos/VotesRepo';
```
Comment (b0b3rt) [inline packages/lesswrong/components/sunshineDashboard/UsersReviewInfoCard.tsx:10]: Unused import? (Can't be used here regardless, since it's a server file)

File: packages/lesswrong/components/sunshineDashboard/UsersReviewInfoCard.tsx [hunk 2]
```diff
@@ -197,7 +200,7 @@ const UsersReviewInfoCard = ({ user, refetch, currentUser, classes }: {
   const showReviewTrigger = reviewTrigger !== 'noReview' && reviewTrigger !== 'alreadyApproved';
   
   if (!userCanDo(currentUser, "posts.moderate.all")) return null
-  
+
```
Comment (b0b3rt) [inline packages/lesswrong/components/sunshineDashboard/UsersReviewInfoCard.tsx:203]: nit: try to avoid accidental whitespace changes

File: packages/lesswrong/components/sunshineDashboard/UsersReviewInfoCard.tsx [hunk 3]
```diff
@@ -7,6 +7,9 @@ import { userCanDo } from '../../lib/vulcan-users/permissions';
 import classNames from 'classnames';
 import { hideScrollBars } from '../../themes/styleUtils';
 import { getReasonForReview } from '../../lib/collections/moderatorActions/helpers';
+import VotesRepo from '../../server/repos/VotesRepo';
+import uniq from 'lodash/uniq';
```
Comment (b0b3rt) [inline packages/lesswrong/components/sunshineDashboard/UsersReviewInfoCard.tsx:11]: Is this getting used?

File: packages/lesswrong/components/sunshineDashboard/UsersReviewInfoCard.tsx [hunk 4]
```diff
@@ -7,6 +7,9 @@ import { userCanDo } from '../../lib/vulcan-users/permissions';
...
+import { RecentKarmaInfo } from '../../server/resolvers/userResolvers';
```
Comment (b0b3rt) [inline packages/lesswrong/components/sunshineDashboard/UsersReviewInfoCard.tsx:12]: Or this?

File: packages/lesswrong/server/repos/VotesRepo.ts [hunk 5]
```diff
@@ -109,4 +116,48 @@ export default class VotesRepo extends AbstractRepo<DbVote> {
...
+    const votes = await this.any(`
+      (
+        SELECT ${voteFields} "Posts"."postedAt"
+        FROM "Votes"
+        JOIN "Posts" on "Posts"._id = "Votes"."documentId"
+        WHERE
+          "Votes"."documentId" in (
+            SELECT _id FROM "Posts" 
+            WHERE
+              "Posts"."userId" = '${userId}'
```
Comment (b0b3rt) [inline packages/lesswrong/server/repos/VotesRepo.ts:135]: This should be passed in as a parameter, not interpolated - SQL injection is a risk here

File: packages/lesswrong/server/repos/VotesRepo.ts [hunk 6]
```diff
@@ -109,4 +116,48 @@ export default class VotesRepo extends AbstractRepo<DbVote> {
...
+            ORDER BY "Posts"."postedAt" DESC
+            LIMIT ${RECENT_CONTENT_COUNT}
```
Comment (b0b3rt) [inline packages/lesswrong/server/repos/VotesRepo.ts:137]: I would either hard-code this, or also pass it in as a parameter.  I think you can reference the same parameter value twice in one query (i.e. use `$2` here and also in the other half of the union)

File: packages/lesswrong/server/repos/VotesRepo.ts [hunk 7]
```diff
@@ -109,4 +116,48 @@ export default class VotesRepo extends AbstractRepo<DbVote> {
...
+              "Comments"."userId" = '${userId}' 
+            ORDER by "Comments"."postedAt" DESC
+            LIMIT ${RECENT_CONTENT_COUNT}
+          )
+          AND
+          "cancelled" IS NOT true
+        ORDER BY "Comments"."postedAt" DESC
+      )
+    `)
+    return votes as RecentVoteInfo[]
```
Comment (b0b3rt) [inline packages/lesswrong/server/repos/VotesRepo.ts:161]: This isn't necessary, you should just be able to `return this.any` and the return type annotation on `getVotesOnRecentContent` will take care of it (since `this.any` already returns a `Promise<any[]>`)

File: packages/lesswrong/server/repos/VotesRepo.ts [hunk 8]
```diff
@@ -33,11 +35,16 @@ export type TagRevisionKarmaChange = KarmaChangeBase & {
   tagId: string,
 }
 
+export interface RecentVoteInfo extends DbVote {
+  postedAt: Date,
```
Comment (b0b3rt) [inline packages/lesswrong/server/repos/VotesRepo.ts:39]: This doesn't seem right; if you're only returning the fields specified in `voteFields`, then this won't have most of the fields on `DbVote` (or `postedAt`)

File: packages/lesswrong/server/resolvers/userResolvers.ts [hunk 9]
```diff
@@ -21,6 +21,82 @@ import { getAnalyticsConnection } from "../analytics/postgresConnection";
...
+  recentPostKarma: number,
+  recentCommentKarma: number,
+  downvoterCount: number, 
+  postDownvoterCount: number,
+  commentDownvoterCount: number,
+}
+
+async function getLatest20contentItems (votes: RecentVoteInfo[]): Promise<RecentVoteInfo[]> {
+  const sortedVotes = votes.sort((a, b) => b.postedAt.getTime() - a.postedAt.getTime())
+  const uniqueVoteDocumentIds = uniq(sortedVotes.map((vote) => vote.documentId))
```
Comment (b0b3rt) [inline packages/lesswrong/server/resolvers/userResolvers.ts:39]: I think you can eliminate the last filter by using `uniqBy` or something?

File: packages/lesswrong/server/resolvers/userResolvers.ts [hunk 10]
```diff
@@ -21,6 +21,82 @@ import { getAnalyticsConnection } from "../analytics/postgresConnection";
...
+export type RecentKarmaInfo = {
+  recentKarma: number, 
+  recentPostKarma: number,
+  recentCommentKarma: number,
+  downvoterCount: number, 
+  postDownvoterCount: number,
+  commentDownvoterCount: number,
+}
+
+async function getLatest20contentItems (votes: RecentVoteInfo[]): Promise<RecentVoteInfo[]> {
```
Comment (b0b3rt) [inline packages/lesswrong/server/resolvers/userResolvers.ts:37]: Are we doing this because getting 10 comments & 10 posts doesn't guarantee 20 items, if the user has less than 10 of one of them?

File: packages/lesswrong/server/resolvers/userResolvers.ts [hunk 11]
```diff
@@ -21,6 +21,82 @@ import { getAnalyticsConnection } from "../analytics/postgresConnection";
...
+  return sortedVotes.filter((vote) => uniqueVoteDocumentsByIdSlice.includes(vote.documentId))
+}
+
+
+export async function getRecentKarmaInfo (userId: string): Promise<RecentKarmaInfo> {
+  const votesRepo = new VotesRepo()
+  const allVotes = await votesRepo.getVotesOnRecentContent(userId)
+  const top20documentVotes = await getLatest20contentItems(allVotes)
+
+  const nonuserIDallVotes = allVotes.filter((vote: RecentVoteInfo) => vote.userId !== userId)
```
Comment (b0b3rt) [inline packages/lesswrong/server/resolvers/userResolvers.ts:50]: Is this filtering out self-votes? If so, we should probably do it in the DB query, if we're always going to filter them out and not use them anywhere

Comment (b0b3rt) [reply packages/lesswrong/server/resolvers/userResolvers.ts:50]: (the name on the variable isn't great :( )

File: packages/lesswrong/server/rateLimits/utils.ts [hunk 12]
```diff
@@ -76,12 +77,36 @@ function getModRateLimitInfo(documents: Array<DbPost|DbComment>, modRateLimitHou
   }
 }
 
-function getAutoRateLimitInfo(user: DbUser, rateLimit: AutoRateLimit, documents: Array<DbPost|DbComment>): RateLimitInfo|null {
-  const { karmaThreshold, downvoteRatio, timeframeUnit, timeframeLength, itemsPerTimeframe, rateLimitMessage, rateLimitType } = rateLimit
+function doAllAutoRateLimitConditionsApply(user: DbUser, rateLimit: AutoRateLimit, recentKarmaInfo: RecentKarmaInfo): boolean {
```
Comment (b0b3rt) [inline packages/lesswrong/server/rateLimits/utils.ts:80]: I am not super clear on what this is doing - it doesn't seem to be checking whether all the rate limit conditions apply... maybe checking that none of them apply?

File: packages/lesswrong/server/repos/VotesRepo.ts (hunk 8 above)
Comment (Raemon) [reply packages/lesswrong/server/repos/VotesRepo.ts:39]: Yeah I was pretty unsure how to handle this correctly, and I was figuring on pairing with you to iron out exactly how to handle the typing here.

File: packages/lesswrong/server/repos/VotesRepo.ts (hunk 6 above)
Comment (Raemon) [reply packages/lesswrong/server/repos/VotesRepo.ts:137]: Is there any way to get rid of SQL injection risk without resorting to the $2 thing? I think the $1 $2 etc pattern makes things a lot less readable and I'd prefer to have some other way of handling this.

File: packages/lesswrong/server/resolvers/userResolvers.ts (hunk 10 above)
Comment (Raemon) [reply packages/lesswrong/server/resolvers/userResolvers.ts:37]: If you don't do this filter, then a user who's ever made one good post will forever have that post be counted as part of their last 20 items, which makes the filter much less useful to detect people who've recently been bad.

File: packages/lesswrong/server/resolvers/userResolvers.ts (hunk 11 above)
Comment (Raemon) [reply packages/lesswrong/server/resolvers/userResolvers.ts:50]: In the current implementation, if you remove the self-votes from the DB query, then you miss out on any comments where only the user self-upvoted them.

BUT, I do acknowledge that relying on their self-upvotes existing to fetch their last 20 comments is pretty brittle. I updated while writing this that we should probably just actually just do a different query, but I figured I would chat with you about what exact query to do instead. Seemed like it required a bit of thinking.

File: packages/lesswrong/server/repos/VotesRepo.ts (hunk 7 above)
Comment (Raemon) [reply packages/lesswrong/server/repos/VotesRepo.ts:161]: It yells at me if I don't include the `as RecentVoteInfo[]`. 

As noted elsewhere, I could use help figuring out exactly how the typing is supposed to work here.

File: packages/lesswrong/server/rateLimits/utils.ts (hunk 12 above)
Comment (Raemon) [reply packages/lesswrong/server/rateLimits/utils.ts:80]: My thinking was "If any of them *don't* apply, then 'it's false that all of them apply'." (although I'd meant to be only conceptually including "conditions the rate limit explicitly included", not "all possible conditions the rate limit could have included")

Comment (Raemon) [reply packages/lesswrong/server/rateLimits/utils.ts:80]: Could rename `doAllListedConditionsApply`?

(maybe we're having different intuitions about whether a condition "applies"? if the karmathreshold is 5, and the user's karma is greater than 5, I'm interpreting that as "the condition does not apply")

**Comment by Raemon** [comment]
It's a bit of a pain to test this, so I'd like to wait on it getting otherwise approved (which can include both code review and "are these quite the right metrics?" reviews)/

Conceptually, it's plausible to me we actually want a "recentDownvoteRatio" score rather than (or at least in addition to) total karma.

**Review by s-cheng — COMMENTED** [review]
(I looked through this a bit since I figured I should be paying more attention to the rate limit changes than I have been recently. I left some comments but didn't do a full review.)

Regarding this:
> We also might want "woah buddy slow down" rate-limits that, say, limit you to one comment per hour if you've recently got lots of unique downvoters and some-but-not-much karma. (though I'd have to see what patterns tend to be common in demon-threads before having a sense of how to calibrate that)

I would guess you could look at the existing demon threads and make a decision based on that?

File: packages/lesswrong/server/rateLimits/utils.ts [hunk 13]
```diff
@@ -252,3 +280,73 @@ export async function rateLimitDateWhenUserNextAbleToComment(user: DbUser, postI
...
+  
+  const downvoters = nonUserIdTop20DocVotes.filter((vote: RecentVoteInfo) => vote.power < 0).map((vote: RecentVoteInfo) => vote.userId)
+  const downvoterCount = uniq(downvoters).length
+  const commentDownvoters = commentVotes.filter((vote: RecentVoteInfo) => vote.power < 0).map((vote: RecentVoteInfo) => vote.userId)
+  const commentDownvoterCount = uniq(commentDownvoters).length
+  const postDownvotes = postVotes.filter((vote: RecentVoteInfo) => vote.power < 0).map((vote: RecentVoteInfo) => vote.userId)
+  const postDownvoterCount = uniq(postDownvotes).length
+
+  // NOTE: the following code is just console logs for sanity checking the above code.
+  // I'm leaving it in for now until I'm more confident that the above code is correct.
```
[the remark this answers is not in the document]
Comment (Raemon) [reply packages/lesswrong/server/rateLimits/utils.ts:320]: Yeah this does make sense as a PR to get some real testing. (fyi we made one really basic rate-limit test last time as a proof of concept).

File: packages/lesswrong/lib/collections/users/fragments.ts [hunk 14]
```diff
@@ -205,6 +205,7 @@ registerFragment(`
     interestedIn
     
     allowDatadogSessionReplay
+    recentKarmaInfo
```
[the remark this answers is not in the document]
Comment (Raemon) [reply packages/lesswrong/lib/collections/users/fragments.ts:208]: ...up, you are correct that I totally forgot to include this on users current. I'm pretty sure I'd have caught this when I went to do testing, but, man, pretty embarrassing.

**Comment by Raemon** [comment]
I think my next action here is to think more through what rate limits I actually want. The comments on "Rate Limits as a moderation tool" have a lot of good points about how autoratelimits might fail or behave weird, and I want to read through that and think a bit more.

https://www.lesswrong.com/posts/HShv7oSW23RWaiFmd/rate-limiting-as-a-mod-tool

I think the basic mechanics of this PR remain roughly the same (fetch the votes on the most recent content, and then do some operations on them and check if some formula is greater than some threshold), but I'm not sure which formula is actually best.

File: packages/lesswrong/lib/collections/users/fragments.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (s-cheng) [reply packages/lesswrong/lib/collections/users/fragments.ts:208]: Sorry I think my comment was confusing, let me rephrase: this line adds `recentKarmaInfo` to `UsersCurrent`, but it doesn't look to me like that is necessary, so I think we should remove this line

[the remark this answers is not in the document]
Comment (Raemon) [reply packages/lesswrong/lib/collections/users/fragments.ts:208]: Yeah I think I'd just assumed you'd need it here in order to properly rate limit the user when they were active, but upon reflection that's actually handled by a different process.

**Comment by oetherington** [comment]
Just double checking if this is actually waiting on my review or not? I wasn't sure since it looked like Sarah already reviewed, and I wasn't sure if the "next action" you mentioned was for this PR or a separate PR?

**Comment by Raemon** [comment]
Oh, I'd been thinking of Sarah's review as satisfying the "EA forum folk keeping tabs on it" review. I have been doing more thinking about this since then and yes was considering the "think about rate limits" as part of this PR. (But, that shouldn't end up affecting EA Forum much since it's just about what rate limits to implement)

**Comment by Raemon** [comment]
That said, some UI updates:

– I have now gotten rid of the ModeratedComments tab in the ModerationDashboard
– I have added a new pseudo-view for ModerationDashboard that shows recentKarma information:

image

File: packages/lesswrong/server/rateLimits/constants.ts [hunk 15]
```diff
@@ -60,6 +61,15 @@ const LW = {
...
       rateLimitMessage: "Users with -30 karma can only post once every two weeks.",
+    },
+    ONE_PER_FOUR_WEEKS_NEGATIVE_60_KARMA: {
+      actionType: "Posts",
+      karmaThreshold: -60,
+      timeframeUnit: 'weeks',
+      timeframeLength: 4,
+      itemsPerTimeframe: 1,
+      rateLimitType: 'lowKarma',
+      rateLimitMessage: `Users with -50 karma on their recent content can only post once every two weeks.`,
```
Comment (darkruby501) [inline packages/lesswrong/server/rateLimits/constants.ts:72]: Is it 50 or 60, these are mismatched. I favor -50

**Comment by Raemon** [comment]
Okay I've now created a column for "auto rate limits that currently affect this user". 

image

File: packages/lesswrong/components/sunshineDashboard/ModerationDashboard.tsx [hunk 16]
```diff
@@ -90,18 +89,7 @@ const styles = (theme: ThemeType): JssStyles => ({
...
-  const commentsWithActions = uniqueComments.map(comment => {
-    const actionsWithoutComment = commentModeratorActions.filter(result => result.comment._id === comment._id).map(({ comment, ...remainingAction }) => remainingAction);
-    return { comment, actions: actionsWithoutComment };
-  });
-
-  return commentsWithActions;
-};
-
-const tabs = new TupleSet(['sunshineNewUsers', 'allUsers', 'moderatedComments'] as const);
+const tabs = new TupleSet(['sunshineNewUsers', 'allUsers', 'recentlyActive', 'moderatedComments'] as const);
```
Comment (b0b3rt) [inline packages/lesswrong/components/sunshineDashboard/ModerationDashboard.tsx:92]: Are we removing the `moderatedComments` tab?

File: packages/lesswrong/lib/rateLimits/utils.ts [hunk 17]
```diff
@@ -0,0 +1,167 @@
+import moment from "moment"
+import { getDownvoteRatio } from "../../components/sunshineDashboard/UsersReviewInfoCard"
+import { AutoRateLimit, RateLimitInfo, RecentKarmaInfo, TimeframeUnitType, UserKarmaInfo, UserRateLimit, rateLimitThresholds } from "./types"
+import { userIsAdmin, userIsMemberOf } from "../vulcan-users"
+import { RecentVoteInfo } from "../../server/repos/VotesRepo"
```
Comment (b0b3rt) [inline packages/lesswrong/lib/rateLimits/utils.ts:5]: This needs to be `import type` for  this import to not blow up if anything on the client imports anything from this file

File: packages/lesswrong/components/sunshineDashboard/ModeratorUserInfo/RecentlyActiveUsers.tsx [hunk 18]
```diff
@@ -0,0 +1,326 @@
...
+const RecentlyActiveUsers = ({ classes }: {
+  classes: ClassesType
+}) => {
+  const { UsersReviewInfoCard, LoadMore, LWTooltip, UsersName, FormatDate, MetaInfo } = Components;
+    
+  const currentUser = useCurrentUser();
+
+  const [expandId, setExpandId] = useState<string|null>(null);
+
+  type sortingType = "lastNotificationsCheck"|"last20Karma"|"downvoters"|"karma"|"lastMonthKarma"|"userSortByRateLimitCount";
```
Comment (b0b3rt) [inline packages/lesswrong/components/sunshineDashboard/ModeratorUserInfo/RecentlyActiveUsers.tsx:118]: Declare this outside the component (& also PascalCase the name)

File: packages/lesswrong/lib/rateLimits/utils.ts [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+    postDownvoterCount: postDownvoterCount ?? 0,
+    commentDownvoterCount: commentDownvoterCount ?? 0,
+    lastMonthDownvoterCount: lastMonthDownvoterCount ?? 0
+  }
+}
+
+export function getRateLimitNames(user: SunshineUsersList, autoRateLimits: AutoRateLimit[]) {
+  const userRateLimits = autoRateLimits.filter(rateLimit => rateLimit.rateLimitType !== "universal")
+
+  function getRateLimitName (rateLimit: AutoRateLimit) {
```
Comment (b0b3rt) [inline packages/lesswrong/lib/rateLimits/utils.ts:154]: Does this need to be defined inside `getRateLimitNames`?

File: packages/lesswrong/components/sunshineDashboard/ModerationDashboard.tsx (hunk 16 above)
Comment (Raemon) [reply packages/lesswrong/components/sunshineDashboard/ModerationDashboard.tsx]: Yes.

File: packages/lesswrong/components/sunshineDashboard/ModeratorUserInfo/RecentlyActiveUsers.tsx [hunk 20]
```diff
@@ -0,0 +1,321 @@
...
+            </td>
+            <td className={classNames(classes.cell, {[classes.selected]: sorting === "userSortByRateLimitCount"})} 
+              onClick={() => setSorting("userSortByRateLimitCount")}>
+              Rate Limits
+            </td>
+            <td></td>
+          </tr>
+        </thead>
+        <tbody>
+          {sortedUsers.map((user, i) => {
```
Comment (b0b3rt) [inline packages/lesswrong/components/sunshineDashboard/ModeratorUserInfo/RecentlyActiveUsers.tsx:251]: Everything inside of this feels like it probably wants to be its own component, but doesn't have to be in this PR

File: packages/lesswrong/lib/rateLimits/utils.ts [hunk 21]
```diff
@@ -0,0 +1,180 @@
+import moment from "moment"
+import { getDownvoteRatio } from "../../components/sunshineDashboard/UsersReviewInfoCard"
```
Comment (b0b3rt) [inline packages/lesswrong/lib/rateLimits/utils.ts:2]: importing from a component to a lib file feels like it's asking for circular dependency issues at some point, but probably fine for now

**Review by b0b3rt — APPROVED** [review]
